### PR TITLE
Jetpack Sync: make jetpack media extraction more consistent

### PR DESCRIPTION
**Pull request by robfelty**
Currently we extract image size info in some cases, but not in all. For example, the `from_slideshow` method will return image width and height, but not `from_gallery`. These changes aim to make this more consistent across all the ways we can extract image information.

Fixes #

## Proposed changes:
This copies much of the logic around extracting image size info to all of the various methods in the `Jetpack_PostImages` class, and adds a number of additional tests for that class and the media extractor class.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
Run the newly added unit tests
```
$ jetpack docker phpunit -- --filter=WP_Test_Jetpack_MediaExtractor
$ jetpack docker phpunit -- --filter=WP_Test_Jetpack_PostImages
```

### CODE REVIEW

**Comment by robfelty** [comment]
There is a small elasticsearch change which is required before this can land D136604-code

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php [hunk 1]
```diff
@@ -450,14 +450,19 @@ protected static function reduce_extracted_images( $images ) {
...
+			}
 			if ( ! empty( $image['alt_text'] ) ) {
-				$ret_images[] = array(
-					'url'      => $image['src'],
-					'alt_text' => $image['alt_text'],
-				);
+				$ret_image['alt_text'] = $image['alt_text'];
 			} else {
-				$ret_images[] = $image['src'];
+				$ret_image = $image['src'];
```
Comment (trakos) [inline projects/plugins/jetpack/_inc/lib/class.media-extractor.php:465]: I think this would overwrite `src_width` and `src_height` when `alt_text` isn't set, is this intentional?

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php [hunk 2]
```diff
@@ -543,14 +549,21 @@ public static function get_images_from_html( $html, $images_already_extracted, $
 			}
 
 			if ( ! in_array( $queryless, $image_list, true ) ) {
```
Comment (trakos) [inline projects/plugins/jetpack/_inc/lib/class.media-extractor.php:553]: Noting that this check will probably work less often now since `$image_list` will store strings less often.

File: projects/plugins/jetpack/class.jetpack-post-images.php [hunk 3]
```diff
@@ -339,7 +356,7 @@ public static function from_thumbnail( $post_id, $width = 200, $height = 200 ) {
 				// If wp_get_attachment_image_src returns false but we know that there should be an image that could be used.
 				// we try a bit harder and user the data that we have.
 				$thumb_post_data = get_post( $thumb );
-				$img_src         = array( $thumb_post_data->guid ?? null, $meta['width'], $meta['height'] );
+				$img_src         = array( $thumb_post_data->guid, $meta['width'], $meta['height'] );
```
Comment (trakos) [inline projects/plugins/jetpack/class.jetpack-post-images.php]: I think this is a left-over from a rebase conflict. Looks like it reverts a recent change from https://github.com/Automattic/jetpack/pull/35264/files. Probably wasn't intentional?

**Review by trakos — COMMENTED** [review]
Tests pass and code looks right to me, though I'm not as familiar with Jetpack codebase.

I have added a few minor notes on a couple of details.

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php [hunk 4]
```diff
@@ -543,14 +549,21 @@ public static function get_images_from_html( $html, $images_already_extracted, $
...
-						'alt_text' => $extracted_image['alt_text'],
-					);
+				$image_to_add = array(
+					'url' => $queryless,
+				);
+				if ( $extract_alt_text ) {
+					if ( ! empty( $extracted_image['alt_text'] ) ) {
+						$image_to_add['alt_text'] = $extracted_image['alt_text'];
+					}
+					if ( ! empty( $extracted_image['src_width'] ) || ! empty( $extracted_image['src_width'] ) ) {
```
Comment (sotirispl) [inline projects/plugins/jetpack/_inc/lib/class.media-extractor.php]: I think the second check should be on `src_height` and not `src_width` again.

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php [hunk 5]
```diff
@@ -543,14 +549,21 @@ public static function get_images_from_html( $html, $images_already_extracted, $
...
+							$image_to_add['src_width']  = $extracted_image['src_width'];
+							$image_to_add['src_height'] = $extracted_image['src_height'];
+					}
 				} else {
-					$image_list[] = $queryless;
+					$image_to_add = $queryless;
```
Comment (sotirispl) [inline projects/plugins/jetpack/_inc/lib/class.media-extractor.php:566]: Shouldn't this also happen when we have `$extract_alt_text=true` but the `$extracted_image['alt_text']` is empty? Is this change intentional?

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php (hunk 1 above)
Comment (robfelty) [reply projects/plugins/jetpack/_inc/lib/class.media-extractor.php:465]: Great question! yes, because the way that `$extract_alt_text` was implemented, when it is set to true, then the results are returned as an array of associative arrays. When it is set to false, then a simple array of urls is returned. I am fairly certain that only ES indexing is using the `$extract_alt_text` option, so I decided to just expand on that, in order to keep other uses backwards compatible.

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php (hunk 2 above)
Comment (robfelty) [reply projects/plugins/jetpack/_inc/lib/class.media-extractor.php:553]: That's a good point. I added another test (and updated one) to make this a bit more explicit. In the case that an image is used once in a gallery and also simply in the content, but with different sizes or alt_text, I think it makes sense to include both. In the case that all the fields are the same, then we should not duplicate. It turns out that we were not uniquing links either. I found a TODO in the code, which I did.

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php (hunk 5 above)
Comment (robfelty) [reply projects/plugins/jetpack/_inc/lib/class.media-extractor.php:566]: Good question. I don't think it should. When `$extract_alt_text` is true, the calling function should be expecting an array of associative arrays. When it is false, it is expecting just an array of strings.

File: projects/plugins/jetpack/_inc/lib/class.media-extractor.php (hunk 4 above)
Comment (robfelty) [reply projects/plugins/jetpack/_inc/lib/class.media-extractor.php]: Thanks. I fixed it!

**Review by sotirispl — APPROVED** [review]
Tests are passing and overall changes look good to me.
